Django 2.0 compatibility

mpauly opened the pull request:
Hi, 
this PR adds Django 2.0 compatibility. 
As the PR https://github.com/django-haystack/django-haystack/pull/1557 did not see any progress recently I implemented everything needed for Django 2.0 compatibility here.

In particular I rewrote the rebuild_index command to strip certain arguments that are not used in the clear_index and update_index commands respectively. This is implemented in a relatively adhoc way, however looking at the amount of changes made to rebuild_index in the past I figured this might be the easiest way to implement this change.

Additionally I added Django 2.0 to the tests and excluded the combination Django2.0 + Python2.7 from the test suite so that the tests still pass. Maybe one should also consider adding python 3.6 to the test suite, however I did not do that yet.

Also note the change I made to `test_haystack/solr_tests/test_solr_backend.py`. This was necessary due to the recent update in pysolr. That update broke the django-haystack tests, as `pysolr.__version__` is now a string instead of a tuple. The fix I implemented should be backwards compatible.

Please let me know if there is anything this MR needs to get merged, cheers,
Martin

acdha commented on test_haystack/solr_tests/test_solr_backend.py:
```diff
@@ -1458,7 +1458,7 @@ def test_boost(self):
 
 
-@unittest.skipIf(pysolr.__version__ < (3, 1, 1), 'content extraction requires pysolr > 3.1.0')
+@unittest.skipIf(tuple(pysolr.__version__.split('.')) < (3, 1, 1), 'content extraction requires pysolr > 3.1.0')
```
This could also use `pysolr.version_info` but that will trigger a deprecation warning on setuptools >= 8.0 since that's no longer a simple tuple.

mpauly replied:
That actually seems to have an additional problem as the parsed version still contains strings instead of integers. 
Either we could do a `pysolr.version_info < parse_version('3.1.1')` with the corresponding import or we just drop that annotation? Seems relatively unlikely that anyone is running the test suite with a version of pysolr from 2013, no?
What do you prefer?

acdha replied:
Dropping it does make sense - if anyone is running tests on an install that old it seems unlikely that they'd have upgraded haystack but frozen pysolr

mpauly replied:
Okay, I dropped the annotation.

mpauly commented:
By the way, thank you @acdha for replying so fast and maintaining django-haystack! 

Anything else that needs updating in this MR?

acdha commented:
I wanted to review it in more detail first but at first glance it looked good

mpauly commented:
@acdha did you already have time to take a look at this MR? any way I can support you in getting this merged?

dani0805 commented:
I can vouch for this, we have been using @mpauly fork for 2 days in our dev branch and it works like a charm.

acdha commented:
@mpauly the main thing is testing – since I’m not using that combination at the moment I haven’t done more than look at the code. I’d like to e.g. confirm test coverage isn’t omitting anything important.

mpauly commented:
@acdha I just added a test to see if ModelSearchIndex properly excludes related fields by default - seems this was not tested before. I also updated the corresponding line in `haystack/indexes.py` as my changes were not working as they should on Django 2.0. Now everything should be working fine and the tests pass on my travis, somehow on yours they don't - see https://travis-ci.org/mpauly/django-haystack. 

Is there any reason why `field.is_relation` was not used before? That seems to be the most generic way to check if a field is a related_field.

mpauly commented:
@acdha I just merged master into this branch (I probably should have rebased - sorry about that)
Additionally I can report that we are using this branch for a few days in our staging environment and at least in our setup it seems to work well.

acdha commented:
Thanks, that's a great confirmation. I'll look into a beta release shortly.

Thanks,
Chris

andrewbenedictwallace commented:
I've been using this fork with Django 2.0 without issues, except that "pip install git+git://github.com/mpauly/django-haystack" downgrades Django to 1.11 (installing from requirements.txt works fine).

I'm assuming this is b/c of setup.py line 15:
install_requires = [
    'Django>=1.8,<1.12',
]

Thanks!

timgraham commented:
I recommend simplifying the patch by dropping support for Django < 1.11 as those versions are unsupported (except for 1.8 for which support ends April 1). See [the note in the 1.11 release notes](https://docs.djangoproject.com/en/dev/releases/1.11/#deprecating-warnings-are-no-longer-loud-by-default).

acdha commented:
Seconding what @timgraham said. I'm content saying that 2.7 will be the last release with support for Django <1.11. We should prune some of the older search engine code, too, as that's also expensive to support but that's a separate ticket.

RabidCicada commented:
There is awesome sauce here. Much thanks.

mpauly commented:
It seems that before I was not running the solr tests correctly, so I had to do more updates to fix some failing tests. Over the weekend I did the following changes:

1. I dropped support for Django < 1.11 - I also tried to cleanup the tox.ini, the travis file and setup.py
2. I added Python 3.6 to the tests in travis
3. I had to replace `get_coords()` by `coords` in multiple places to fix some tests
4. I updated some of the test configuration to work properly with Django 2.0

There might potentially be bugs in some functionality related to the geo lookups, however tests are passing with Django 1.11 and 2.0. We are also running this fork in production with Django 2.0 and have not experienced any problems yet.

timgraham commented on .travis.yml:
```diff
@@ -58,17 +59,18 @@ script:
-        - DJANGO_VERSION=">=1.8,<1.9" VERSION_ES=">=2.0.0,<3.0.0"
-        - DJANGO_VERSION=">=1.9,<1.10" VERSION_ES=">=2.0.0,<3.0.0"
-        - DJANGO_VERSION=">=1.10,<1.11" VERSION_ES=">=2.0.0,<3.0.0"
+        - DJANGO_VERSION=">=1.11,<1.12" VERSION_ES=">=1.0.0,<2.0.0"
```
<1.12 -> <2.0

timgraham commented on .travis.yml:
```diff
@@ -58,17 +59,18 @@ script:
 matrix:
     allow_failures:
         - python: 'pypy'
+        - python: 2.7
```
These builds should be excluded rather than "allowed_failures".

timgraham commented on haystack/management/commands/rebuild_index.py:
```diff
@@ -33,5 +34,11 @@ def add_arguments(self, parser):
     def handle(self, **options):
-        call_command('clear_index', **options)
-        call_command('update_index', **options)
+        clear_options = copy.copy(options)
```
`options.copy()`?

timgraham commented on test_haystack/solr_tests/test_admin.py:
```diff
@@ -2,11 +2,12 @@
 
 from __future__ import absolute_import, division, print_function, unicode_literals
 
+import django
```
unused

timgraham commented on setup.py:
```diff
@@ -12,7 +12,7 @@
 
 install_requires = [
-    'Django>=1.8,<1.12',
+    'Django>=1.11,<2.1',
```
I wouldn't specify an upper bound -- it might be that this will work fine with Django 2.1 without any changes.

timgraham commented on test_haystack/core/models.py:
```diff
@@ -17,7 +17,7 @@ class MockModel(models.Model):
     foo = models.CharField(max_length=255, blank=True)
     pub_date = models.DateTimeField(default=datetime.datetime.now)
-    tag = models.ForeignKey(MockTag)
+    tag = models.ForeignKey(MockTag, on_delete=models.CASCADE)
```
You can omit `on_delete=` and pass it as a positional argument... that's what we've done in Django's test models anyway.

timgraham commented on test_haystack/test_app_loading.py:
```diff
@@ -3,7 +3,8 @@
 from types import GeneratorType, ModuleType
 
-from django.core.urlresolvers import reverse
+import django
```
unused

timgraham commented on test_haystack/test_views.py:
```diff
@@ -5,11 +5,12 @@
 import time
 from threading import Thread
 
+import django
```
unused

timgraham commented on tox.ini:
```diff
@@ -1,48 +1,32 @@
+[django1.11]
 deps =
-    Django>=1.8,<1.9
+    Django>=1.11,<1.12
```
<2.0

timgraham commented on test_haystack/test_management_commands.py:
```diff
@@ -87,25 +90,30 @@ def test_rebuild_index_nocommit(self, *mocks):
         """
         from haystack.management.commands.rebuild_index import Command
 
+        update_mock.return_value = ''
```
You can put `return_value=''` in the `@patch` calls.

timgraham commented on test_haystack/test_indexes.py:
```diff
@@ -623,6 +629,7 @@ def test_excludes(self):
         self.assertTrue(isinstance(self.emsi.fields['pub_date'], indexes.DateTimeField))
         self.assertTrue('text' in self.emsi.fields)
         self.assertTrue(isinstance(self.emsi.fields['text'], indexes.CharField))
+        self.assertFalse('related_models' in self.m2mmsi.fields)
```
`assertNotIn`

timgraham commented on test_haystack/test_management_commands.py:
```diff
@@ -69,6 +69,9 @@ def test_rebuild_index_using(self, m1, m2):
     @patch('haystack.management.commands.update_index.Command.handle')
     @patch('haystack.management.commands.clear_index.Command.handle')
     def test_rebuild_index(self, mock_handle_clear, mock_handle_update):
+        mock_handle_clear.return_value = ''
```
You can put `return_value=''` in the `@patch` calls.

timgraham commented on test_haystack/test_management_commands.py:
```diff
@@ -87,25 +90,30 @@ def test_rebuild_index_nocommit(self, *mocks):
+        args, kwargs = clear_mock.call_args
+
+        self.assertIn('interactive', kwargs)
+        self.assertEqual(False, kwargs['interactive'])
```
for boolean comparisons: `self.assertIs(kwargs['interactive'], False)`

mpauly commented:
@timgraham Thanks for the comments, all of them should be implemented now.

webtweakers commented:
I can confirm that @mpauly's fork works fine with Django 2.0. Tested rebuild_index command, searching, faceting and autocomplete in combination with ElasticSearch. No problems here.

michaelmior commented:
In my case, django-haystack 2.7.0 works fine with Django 2 without any changes. Would be great to have this explucitly supported though.

nadimtuhin commented:
waiting for d2 support :D

JonLevischi commented:
Any idea when this branch will be officially committed to the master package?
Thanks

acdha commented:
It's waiting for more people to test it. If you have time to volunteer in
that regard it'll happen faster.

RabidCicada commented:
@acdha WRT more people testing, what are we looking for?  5, 10?  I don't have a production system.  But I'm happy to test with my rinky dink setup, and vouch if it works for me.  How many vouchsafes are we short?  I don't think people, myself, included like fuzzy "more people" statements.

acdha commented:
I was thinking more different combinations: i.e. I don't use ElasticSearch
with Haystack much so at the very least I think it'd make sense to make
sure someone has tested it with at least Solr, ES, and Whoosh.

RabidCicada commented:
Ahh...very good clarification.  Thanks.  I wont be able to help much on that front :)

nadimtuhin commented:
tested with django 2 & elasticsearch on aws, everything worked fine

JonLevischi commented:
Apparently it will work with Whoosh on AWS as well.

mpauly commented:
We are running Xapian-haystack, that works well with this branch

acdha commented:
This is now in https://pypi.python.org/pypi/django-haystack/2.8.0
